Financial-Times/dotcom-page-kit
# Refactor page-kit-cli into a Webpack configuration builder

# This PR is part 1 of 2, and should not be reviewed in isolation. Part 2: #778

together with #778, this closes #528.

**summary of changes**:

- removes all CLI-related parts of `dotcom-page-kit-cli`
- exposes `getWebpackConfig` as its main export
- renames it to `dotcom-build-webpack-config`

**why we're doing this**: in FT.com we have a history of building custom tooling and letting it rot for several years leaving nobody able to support it. `page-kit-cli` has a complex architecture, and while it was originally intended to support tasks other than `build`, they never materialised, leaving a poorly-understood wrapper around `webpack-cli` whose only benefit is to allow people to customise the configuration in ways we can't support.

tl;dr less custom stuff, more developer confidence

## how to test this PR

sort answer: don't. long answer: test #778 instead. if it works there it'll work here; if it doesn't work there we probably can't fix it here. you can run webpack in the examples and check the output there if you like but testing #778 will be simpler.

## how to review this PR

- the main entry point for the new non-CLI package is [`packages/dotcom-build-webpack-config/src/operations/getWebpackConfig.ts`](https://github.com/Financial-Times/dotcom-page-kit/pull/777/files#diff-ddfed7352f76cca6c205632a99cba2b3), but this hasn't actually changed very much [from `master`](https://github.com/Financial-Times/dotcom-page-kit/blob/b888454/packages/dotcom-page-kit-cli/src/operations/getWebpackConfig.ts). the main things that _have_ changed there are formatting and names of things (e.g. since it's no longer a CLI it doesn't make sense to have a class called `CliContext`)
- have a look at what's been removed between `dotcom-page-kit-cli` and `dotcom-build-webpack-config`, it should be all CLI-related stuff, and everything should still fit together. note that Github hasn't picked up on some moved (not deleted) files so check filenames
- the changes in the other `dotcom-build-` plugins are to do with removing access to settings from `dotcom-page-kit-cli` which we're mostly making unconfigurable (it doesn't make sense to let people change the `babel-preset-env` settings, that'll just cause different bundle contents to other apps for no reason)
- the examples have been migrated from `page-kit-cli` to the `getWebpackConfig`-based Webpack config format. they should all work.
- because this is an intermediate step, some of the documentation is incomplete, as it's pointless documenting something that's immediately going to be refactored in the next PR without ever reaching users. please review the finished documentation in #778.
